src/inmanta/references.py

```diff
@@ -0,0 +1,450 @@
+        self._reference_value_cached: bool = False
+
+    @abc.abstractmethod
+    def resolve(self) -> T:
+        """This method resolves the reference and returns the object that it refers to"""
+
+    def get(self) -> T:
+        """Get the value. If we have already resolved it a cached value is returned, otherwise resolve() is called"""
+        if not self._reference_value_cached:
+            self._reference_value = self.resolve()
+            self._reference_value_cached = True
+
+        return self._reference_value
+
+    def serialize(self) -> ReferenceModel:
+        """Emit the correct pydantic objects to serialize the reference in the exporter."""
+        if not self._model:
+            argument_id, arguments = self.serialize_arguments()
+            self._model = ReferenceModel(id=argument_id, type=self.type, args=arguments)
+
+        assert isinstance(self._model, ReferenceModel)
+        return self._model
+
+
+class DataclassReference[T: DataclassProtocol](Reference[T], metaclass=DataclassRefeferenMeta):
+    @classmethod
+    def get_dataclass_type(cls) -> type[DataclassProtocol]:
+        """Get the dataclass type that the reference points to"""
+        for g in cls.__orig_bases__:  # type: ignore
+            if typing_inspect.is_generic_type(g):
+                for arg in typing.get_args(g):
+                    if dataclasses.is_dataclass(arg) and isinstance(arg, type):
+                        return arg
+
+        raise TypeError("A dataclass reference requires a typevar that is bound to the dataclass it references.")
+
+    def __getattr__(self, name: str) -> object:
+        """If our reference is a subclass of Value (dataclass), we return a reference to the
+        attribute instead of the attribute itself.
+
+        :param name: The name of the attribute to fetch
+        :return: A reference to the attribute
+        """
+        dc_type = self.get_dataclass_type()
+        if name == DATACLASS_FIELDS:
+            return getattr(dc_type, DATACLASS_FIELDS)
+
+        fields = {f.name: f for f in dataclasses.fields(dc_type)}
+        if name in fields:
+            return AttributeReference(
+                reference=self,
+                attribute_name=name,
+                attribute_type=typing.cast(type[PrimitiveTypes], fields[name].type),
+            )
+
+        raise AttributeError(name=name, obj=self)
+
+
+# TODO: we need to make sure that the executor knows it should load the mutator and executor code before running the handler
+#       of a resource that uses references.
+
+
+class reference:
+    """This decorator register a reference under a specific name"""
+
+    # It is not allowed to use T in a class var so we cannot use T here
+    _reference_classes: typing.ClassVar[dict[str, type[Reference[RefValue]]]] = {}
+
+    def __init__(self, name: str) -> None:
+        """
+        :param name: This name is used to indicate the type of the reference
+        """
+        self.name = name
```

Comment:
I think this will need to be changed somewhat with what we discussed earlier today. i.e. we'll want to accept a Python type and convert it with `to_dsl_type`. Would be even nicer if we can derive the Python type from the type subscription (i.e. the value for the generic type var) of the `Reference`, but I'm not sure how feasible that would be tbh.